### PR TITLE
Pydantic evented dataclass

### LINKED ISSUE #2009: Add type checking and parsing for our models
## 🚀 Feature
Having type checking for our core model files could be very beneficial. 

As has been discussed here https://github.com/napari/napari/issues/1490#issuecomment-735314450 and was originally proposed by @tlambert03 in https://github.com/napari/napari/issues/1436#issue-655404990 we could [pydantic](https://pydantic-docs.helpmanual.io/) to our evented dataclasses to do this.

As I understand it, all of our models would have to inherit from their [BaseModel](https://pydantic-docs.helpmanual.io/usage/models/) or we could try and get their [dataclass](https://pydantic-docs.helpmanual.io/usage/dataclasses/) working with ours. I think the dataclass option sounds like it might be nicer,  but requires more investigation.

If we went with the `BaseModel` approach we'd get methods like `dict()` for free, which might be duplicative of the `asdict()` method we recently added in #1966.

We might also be able to use [validators](https://pydantic-docs.helpmanual.io/usage/validators/) to check when there is a complex relationships between objects, like we've got in a couple of the [Dims._on_*_set](https://github.com/napari/napari/blob/master/napari/components/dims.py#L183) methods in the dims component, potentially simplifying that file.

It might be nice to look into this before we release 0.4.3 which contains a lot more evented dataclass code, in case we need to make some breaking to what we've recently merged.

Curious what @jni @tlambert03 @Czaki @kevinyamauchi and others think about this?

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This PR is a proof of concept of making an evented pydantic dataclass. As mentioned in #2009 in I dropped the `Properties` functionality and decided to just make a new decorator `evented` that can decorate any dataclass, pydantic or not. I've preserved through the `asdict` and `update` methods too.

I'd call this very much a proof of concept. What's nice though is that it if we like this we could gradually role this change out to all our evented dataclasses. Note these are a lot less powerful than the other ones, there is no special setter, no special getter, and no private `_on_*_set` method, but for many of our evented dataclasses we might not need the additional functionality.

Very curious what @jni @tlambert03 and others think. A major motivation for this (as discussed in #2009) is that it now makes things like interpolation between states possible, serialization is much cleaner, and to quote @tlambert03 who originall proposed this in https://github.com/napari/napari/issues/1436#issuecomment-657281399

> If we implemented this thoroughly with pydantic, we'd have a rock solid model that was nearly guaranteed to be consistent.

## Type of change

- [x] Refactor (non-breaking change which fixes an issue)

# References

Closes #2009

### CODE REVIEW

**Comment by sofroniewn** [comment]
Next big question is if I can refactor `Dims` to use the pydantic validators https://pydantic-docs.helpmanual.io/usage/validators/

**Comment by sofroniewn** [comment]
@kevinyamauchi you might be interested in this too, especially in the context of #1940. Also note the pydantic has what looks like a very nice color type we might want to use!! https://pydantic-docs.helpmanual.io/usage/types/#color-type

**Comment by tlambert03** [comment]
certainly intriguing!  Sorry I didn't get around to commenting on https://github.com/napari/napari/issues/2009#issuecomment-748677780 yet... but yeah, as you've shown here, the pydantic validator should mostly replace the need for `Property` in the setter.  A couple things that I'm thinking here:
- how will this work with things like our StringEnums, wehere we tend to have a pattern of using enums internally but presenting plain strings externally?  that was one thing that the `Property` type was enabling, though I'd personally be ok with just making them always be enums (including on `get`).  I suspect the "pydantic way" would say that that was a bad pattern to begin with (type ambiguity) ... but it's something we should consider since we use it multiple places.
- a (minor) motivation on the `properties` was making it so that the interactive documentation of an attribute would retain the documentation of our original model style, so that typing `viewer.theme?` for instance would give you something better than the documentation for python's builtin `str` type.  That too though, I'm ok with losing... (in fact, though [there is a test for it](https://github.com/napari/napari/blob/e61e4fc85b94597b32952599e565330e66d42f1f/napari/utils/events/_tests/test_dataclass.py#L78-L81), on a quick check I'm not sure it's working for all classes at the moment?)
- I suspect you removed the `_on_<name>_set` option in your implementation here just for the sake of simplicity while testing (or maybe because of the private attribute issue you mentioned with pydantic?) ... but if not, do you have a though on how you might conditionally prevent event emission, or allow side effects here?
- can you give me an example on what you mean by this easing interpolation between states?

In general I feel pretty positive about this!  I do think we'd want to do some simple benchmarks to make sure that pydantic isn't adding too much overhead... but barring that, I've always liked pydantic and I'm glad you're exploring this!  I also _like_ the dual decorator approach here.

lemme know what I can help with and if there are major outstanding questions

**Comment by sofroniewn** [comment]
> I also like the dual decorator approach here.

Nice, glad you like that!

> how will this work with things like our StringEnums

Hmm yeah, I havn't looked into that one yet, but good flag.

> interactive documentation of an attribute would retain the documentation of our original model style

Ah right, i wonder if there is another way to achieve that while still using pydantic.

> I suspect you removed the _on_<name>_set option in your implementation here just for the sake of simplicity while testing (or maybe because of the private attribute issue you mentioned with pydantic?) ... but if not, do you have a though on how you might conditionally prevent event emission, or allow side effects here?

I have removed, it could be added back in I think though. I want to see how far I can get with validator side effects, but i might not be able to do everything. I think the `Dims` model will be a big test of that. Blocking event emission might be tricky without it, but I'm probably ok with either using the external blocker if you want to block all emission or something failing validation and no event getting emitted, but we'll see.

> can you give me an example on what you mean by this easing interpolation between states?

I want to do things like interpolate between two `int` using integer steps and two `float` using `float` steps. Without the type coercion we could have `camera.zoom` which should be a float, but then someone sets it to an int, and then sets it to an int again - now I have two `int` values of zoom and so I by mistake just move in integer steps leading to a big discretization artifact.

----------

> lemme know what I can help with and if there are major outstanding questions

One major outstanding question is how to deal with properties that are numpy arrays. It looks like it is possible with `BaseModel`, see https://github.com/samuelcolvin/pydantic/issues/380#issuecomment-620378743 or https://github.com/samuelcolvin/pydantic/issues/380#issuecomment-736135687, but I'm running into problems using this on a dataclass due to `==` checking and a need for a `.all()` call.

I can use a `default_factory` to make some progress but I'm still running into errors
```python
class _ArrayMeta(type):
    def __getitem__(self, t):
        return type('Array', (Array,), {'__dtype__': t})


class Array(np.ndarray, metaclass=_ArrayMeta):
    @classmethod
    def __get_validators__(cls):
        yield cls.validate_type

    @classmethod
    def validate_type(cls, val):
        dtype = getattr(cls, '__dtype__', None)
        if isinstance(dtype, tuple):
            dtype, shape = dtype
        else:
            shape = tuple()

        result = np.array(val, dtype=dtype, copy=False, ndmin=len(shape))
        assert not shape or len(shape) == len(
            result.shape
        )  # ndmin guarantees this

        if any(
            (shape[i] != -1 and shape[i] != result.shape[i])
            for i in range(len(shape))
        ):
            result = result.reshape(shape)
        return result


def make_default_color_array():
    return np.array([1, 1, 1, 1])


@evented
@dataclass(config=PydanticConfig)
class Axes:
    """Axes indicating world coordinate origin and orientation.

    Attributes
    ----------
    visible : bool
        If axes are visible or not.
    labels : bool
        If axes labels are visible or not. Not the actual
        axes labels are stored in `viewer.dims.axes_labels`.
    colored : bool
        If axes are colored or not. If colored then default
        coloring is x=cyan, y=yellow, z=magenta. If not
        colored than axes are the color opposite of
        the canvas background.
    dashed : bool
        If axes are dashed or not. If not dashed then
        all the axes are solid. If dashed then x=solid,
        y=dashed, z=dotted.
    arrows : bool
        If axes have arrowheads or not.
    background_color : np.ndarray
        Background color of canvas. If axes are not colored
        then they have the color opposite of this color.
    """

    visible: bool = False
    labels: bool = True
    colored: bool = True
    dashed: bool = False
    arrows: bool = True
    background_color: Array[float, (-1, 4)] = field(default_factory=make_default_color_array)
```

```python-traceback
  File "/Users/nsofroniew/GitHub/napari/napari/components/axes.py", line 47, in <module>
    @dataclass(config=PydanticConfig)
  File "pydantic/dataclasses.py", line 224, in pydantic.dataclasses.dataclass.wrap
    'default_factory',
  File "pydantic/dataclasses.py", line 160, in pydantic.dataclasses._process_class
    class _HAS_DEFAULT_FACTORY_CLASS:
  File "pydantic/main.py", line 931, in pydantic.main.create_model
  File "pydantic/main.py", line 262, in pydantic.main.ModelMetaclass.__new__
  File "pydantic/fields.py", line 315, in pydantic.fields.ModelField.infer
  File "pydantic/fields.py", line 284, in pydantic.fields.ModelField.__init__
  File "pydantic/fields.py", line 349, in pydantic.fields.ModelField.prepare
  File "pydantic/fields.py", line 378, in pydantic.fields.ModelField._set_default_and_type
  File "pydantic/fields.py", line 287, in pydantic.fields.ModelField.get_default
  File "pydantic/utils.py", line 621, in pydantic.utils.smart_deepcopy
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

There are a number of interesting threads in pydantic about their `dict` method and serialization including of array data. I'm not quite sure what we need yet, but I am wondering if when this is done we'll want to use some of their methods instead of our own `asdict` / `update`, and I'm still not sure about how we're handling the `==` check to block event emission too, compared to things they may be doing. This all might be of interest to @Czaki too

#### tl;dr any help getting numpy arrays working on a pydantic dataclass right now would be appreciated

**Comment by Czaki** [comment]
> I'm still not sure about how we're handling the `==` check to block event emission too, compared to things they may be doing. This all might be of interest to @Czaki too

We are using dict to keep comparison method for fields which cannot be compared with `==`

**Comment by sofroniewn** [comment]
Ok, I've made some pretty major progress figuring out how to get numpy arrays working with pydantic and I have to say I love it. One thing though is that i'm having much better success using the pydantic BaseModel rather than the pydantic dataclass, it's more fully featured, and seems more widely used. I've gotten the evented decorator working with it. The code right now in this PR is all over the place, I'll clean it up when I get the chance, but curious @tlambert03 how you think about the evented decorator + `BaseModel` combination? Do you see any downsides or should I press on?

I'm particularly excited that `BaseModel` provides it's own `dict` and `json` [methods](https://pydantic-docs.helpmanual.io/usage/exporting_models/#modeldict) and you can add custom encoders that will do things like convert numpy to lists if you want etc etc.

If we go down this pass it would be nice to do before the `0.4.3` release as we'd break some things like https://github.com/napari/napari/pull/1966 which haven't been released yet. I do see this as a major choice though that we want to take our time with!

**Comment by Czaki** [comment]
There are any plans to do a performance comparison before merging these PR?

**Comment by sofroniewn** [comment]
Update: This exploration continues to go really well. I'm pretty excited about using the pydantic [BaseModel](https://pydantic-docs.helpmanual.io/usage/models/). I've figure out how to do numpy arrays with different types/ shapes. 

I was able to get everything we needed for the dims validation with the [root_validator](https://pydantic-docs.helpmanual.io/usage/validators/#root-validators)

I really like their `dict` and `json` methods and can get them to play nice with numpy. They include nice arguments like [exclude](https://pydantic-docs.helpmanual.io/usage/exporting_models/#modeldict) which we would have needed to add to our `asdict` method at some point.

> how will this work with things like our StringEnums, 

It's a bit awkward, and I personally would advocate we switch over to a [normal enum pattern which pydantic can use](https://pydantic-docs.helpmanual.io/usage/types/#enums-and-choices). Note that people can still set things with strings `cursor.style = 'circle'` but `cursor.style` will now be an enum, which I think will be much less confusing than having the private property be the enum.

We also might want to serious consider using some of their [color types](https://pydantic-docs.helpmanual.io/usage/types/#color-type) when we just need single colors - not sure if we'd switch over ColorArrays @kevinyamauchi too.

I think we should then consider using [pydantic for our settings management](https://pydantic-docs.helpmanual.io/usage/settings/) too, for example #1183.

> There are any plans to do a performance comparison before merging these PR?

Yes, we'll do some benchmarks, so far under "normal" usage I havn't seen anything concerning.

--------

The code in this PR is still all over the place, but I'll probably clean it up soon, and make this PR complete. Would be great to hear any additional concerns/ feedback from people before then. I might even close this PR and start a new one. It depends.

**Comment by sofroniewn** [comment]
This PR will be superseded by #2042, which will use the pydantic `BaseModel`. Discussion can continue over there
